Automattic/newspack-blocks
# Block Patterns Infrastructure

### All Submissions:






### Changes proposed in this Pull Request:



Introduces UI and infrastructure for "Block Patterns:" preset groupings of blocks that can be inserted into posts. Similar to https://github.com/Automattic/newspack-blocks/pull/371, but this PR contains no actual patterns, only the infrastructure. 

Block patterns are defined by using the `newspack_blocks_patterns` filter. In the editor, if there are any patterns defined, there will be an icon to the right of the Settings icons. Clicking reveals the Newspack Patterns sidebar, which contains all defined patterns. Clicking the pattern inserts the blocks into the editor.

Example implementation:

Updated: `icon` parameter renamed to `preview_image`.
```
add_filter(
	'newspack_blocks_patterns',
	function( $patterns, $post_type ) {
		if ( in_array( $post_type, [ 'post', 'page' ], true ) ) {
			$patterns[] = [
				'category' => __( 'Homepage Posts', 'newspack-blocks' ),
				'content'  => '<h2>Hello World</h2>',
				'preview_image'     => '/path/to/icon/image',
				'title'    => __( 'Test Pattern', 'newspack-blocks' ),
			];
		}
		return $patterns;
	},
	10,
	2
);
```

The `$post_type` parameter can be used to register patterns only for that post type. `content` can be any valid block markup. To test, simply create some blocks, switch to Code Editor, and copy paste. Patterns will be grouped by Category. Categories will be listed alphabetically. `icon` is a path to an icon to use for the pattern.

### How to test the changes in this Pull Request:

1. Initially verify that the patterns icon is NOT shown in the editor. It should only be shown if there are registered patterns available.
2. Register one or more patterns using code like the example above. For real world examples, try the ones in https://github.com/Automattic/newspack-blocks/pull/371. Icons are available in that PR as well.
3. Return to the Editor. Verify the icon is present. Click the icon, verify the UI. Click on any of the patterns and verify the blocks are inserted into the post.

### Other information:
